### PR TITLE
Update for Telegram Bots 2.0

### PR DESCRIPTION
**Pull request by tjhorner**
# Telegram Bots 2.0!

I've added the new methods from the updated Bots API in Telegram. This includes all of the methods for kicking/banning users as well.

The APIs for kicking and unbanning chat members don't seem to work right now because it says the bots must be admins in the groups, but there's no way to make bots admins in groups quite yet. (unless I'm missing something...)

EDIT: It seems that the developers are aware that users cannot make bots admins and are fixing it. Personally, I think they're trying a slow rollout.

### Methods added

- [`kickChatMember`](https://core.telegram.org/bots/api#kickchatmember)
- [`unbanChatMember`](https://core.telegram.org/bots/api#unbanchatmember)
- [`answerCallbackQuery`](https://core.telegram.org/bots/api#answercallbackquery)
- [`editMessageText`](https://core.telegram.org/bots/api#editmessagetext)
- [`editMessageCaption`](https://core.telegram.org/bots/api#editmessagecaption)
- [`editMessageReplyMarkup`](https://core.telegram.org/bots/api#editmessagereplymarkup)

### CODE REVIEW

**Comment by yagop** [comment]
Nice, you were really quickly. :clap: :clap: :clap: 

Some test are needed. Maybe in some functions is almost impossible (the project doesn't have a mocked server) to test but others are.

**Comment by tjhorner** [comment]
Sure, I'll add some tests when I get the time.

**Comment by fcuellom** [comment]
Very good work!!!
I've done something similar, but I added a parameter (show_alert) to the " answerCallbackQuery " function:

`TelegramBot.prototype.answerCallbackQuery = function (callback_query_id, text, show_alert) {
  var form = {};
  form.callback_query_id = callback_query_id;
  form.text = text;
    form.show_alert = show_alert;
  return this._request('answerCallbackQuery', {form: form});
};`

Are you planning to pull the code?

Thanks in advance!

**Comment by tim427** [comment]
Hi @tjhorner ,

I've tried your pull-request, but receive the following error;

```
/home/tim/telegram-bot/node_modules/node-telegram-bot-api/src/telegram.js:17
class TelegramBot extends EventEmitter {
^^^^^

SyntaxError: Block-scoped declarations (let, const, function, class) not yet supported outside strict mode
    at exports.runInThisContext (vm.js:53:16)
    at Module._compile (module.js:373:25)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Module.require (module.js:353:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (/home/tim/telegram-bot/node_modules/node-telegram-bot-api/index.js:1:80)
    at Module._compile (module.js:409:26)
    at Object.Module._extensions..js (module.js:416:10)
error: Forever detected script exited with code: 1
```

Which node-version are you using?

**Comment by tjhorner** [comment]
Seems to work for me, @tim427. It also doesn't seem related to my pull request, but I'm on `v4.2.2`.

**Comment by tjhorner** [comment]
Okay,  you can now make bots admins in regular groups (not supergroups!) and kick chat members. You still can't unban them since you really can't ban users in regular groups. I'll push the tests soon.

**Comment by icdevin** [comment]
Doesn't it need to be modified in order to properly receive POSTs with `callback query` data?

**Comment by tjhorner** [comment]
@icdevin You are correct, if you'd like to add that go ahead, but this PR isn't really at the top of my priority list. Also bots can be admined in supergroups now. I don't think I'll actually be able to write any of the tests soon, so if someone wants to take over that would be nice...

**Comment by icdevin** [comment]
Sure. I'll write the tests. Seems to be a pretty popular request. I'll need some time to do the group stuff as I haven't actually used it before but shouldn't be too difficult, I imagine. Thanks for your work thus far, @tjhorner.

**Comment by tjhorner** [comment]
That could work. If you fork my repo and checkout the `bots-2.0` branch, it should preserve my changes and add yours to this PR.

**Comment by icdevin** [comment]
Does anyone else get a lot of timeouts when running the normal tests? I find a majority of them taking over the default 2000ms threshold for failure. When I upped it to 5000ms, all passed with the exception of two sendVideo tests. @yagop?

**Comment by yagop** [comment]
Just depends on you network speed and how bad are the Telegram servers. Currently I got 964ms on sendVideo, if you want you can pass `--timeout` to mocha, a PR is welcomed.

**Comment by SemihAkar** [comment]
when will the merge happen?

**Comment by icdevin** [comment]
I'm gonna try to finish this up in the next couple days. Sorry guys, been on vacation 😄!

**Comment by SemihAkar** [comment]
@icdevin thanks, we wait for it

**Comment by icdevin** [comment]
@yagop: Removed the extraneous done() calls.

**Comment by zigzag-way** [comment]
I updated my Node to the latest one:

node -p process.versions.v8
5.0.71.35
node -v
v6.1.0
... but still get errors:
node_modules\node-telegram-bot-api\src\telegram.js:20
static messageTypes = [
^
SyntaxError: Unexpected token =

Im sure its because of ES6... but dont understant how to fix it.

**Comment by phillipadsmith** [comment]
@zigzag-way Does this question have anything to do with this issue? If not, could you please open a new issue that's specific to your problem?

**Comment by zigzag-way** [comment]
I posted it here because I have the same problem as one of the guys posted here.

**Comment by zigzag-way** [comment]
And this problem is related with this Bot2.0 fork because it`s rewritten in ES6

**Comment by derhuerst** [comment]
@zigzag-way You could create an issue on the fork.

File: src/telegram.js [hunk 1]
```diff
@@ -455,6 +459,119 @@ class TelegramBot extends EventEmitter {
...
+   * bot or via the bot (for inline bots). On success, the
+   * edited Message is returned.
+   *
+   * Note that you must provide one of chat_id, message_id, or
+   * inline_message_id in your request.
+   *
+   * @param  {String} caption  New caption of the message
+   * @param  {Object} [options] Additional Telegram query options (provide either one of chat_id, message_id, or inline_message_id here)
+   * @return {Promise}
+   * @see https://core.telegram.org/bots/api#editmessagetext
```
Comment (smmoosavi) [inline src/telegram.js]: https://core.telegram.org/bots/api#editmessagecaption

File: src/telegram.js [hunk 2]
```diff
@@ -455,6 +459,119 @@ class TelegramBot extends EventEmitter {
...
+   * sent by the bot or via the bot (for inline bots).
+   * On success, the edited Message is returned.
+   *
+   * Note that you must provide one of chat_id, message_id, or
+   * inline_message_id in your request.
+   *
+   * @param  {Object} replyMarkup  A JSON-serialized object for an inline keyboard.
+   * @param  {Object} [options] Additional Telegram query options (provide either one of chat_id, message_id, or inline_message_id here)
+   * @return {Promise}
+   * @see https://core.telegram.org/bots/api#editmessagetext
```
Comment (smmoosavi) [inline src/telegram.js]: https://core.telegram.org/bots/api#editmessagereplymarkup

**Comment by tjhorner** [comment]
As it seems many people are having problems with this PR, I am going to make a new fork of the latest version of the library and try to implement it with as much compatibility as possible. I'll also add the listeners for callback requests and some of the functions I missed. I will most likely have time sometime this week.

**Comment by SemihAkar** [comment]
When will the merge happen?

**Comment by Kaos1337** [comment]
For who needs a not ES6 version with bot 2.0, here there is what I use: https://github.com/Kaos1337/node-telegram-bot-api

**Comment by yagop** [comment]
Merged and published 0.22.0.

I'm really really sorry for not doing it before but I haven't too much time and has lost a bit of motivation.
Anyway it's merged now, enjoy and please forgive me if you can.

**Comment by tjhorner** [comment]
No worries @yagop! I've just been using my local version for my projects for now.

File: src/telegram.js [hunk 3]
```diff
@@ -110,6 +111,9 @@ class TelegramBot extends EventEmitter {
     } else if (chosenInlineResult) {
       debug('Process Update chosen_inline_result %j', chosenInlineResult);
       this.emit('chosen_inline_result', chosenInlineResult);
+    } else if (callbackQuery) {
+      debug('Process Update callback_query %j', callbackQuery);
+      this.emit('callback_query', callbackQuery);
```
Comment (smmoosavi) [inline src/telegram.js]: need mention in readme#Events

**Pull request merged**